src/java.base/share/classes/java/util/random/RandomGenerator.java

```diff
@@ -252,6 +253,169 @@ default DoubleStream doubles(long streamSize, double randomNumberOrigin,
+             * sparse products k eps, that is, fewer doubles in I.
+             * delta is therefore the best value to ensure the largest number
+             * of equidistant doubles in the interval I.
+             *
+             * left / delta is an exact double and an exact integer with
+             *      -2^P <= left / delta <= 0
+             * Thus, kl is computed exactly.
+             *
+             * Mathematically,
+             *      kr = ceil(right / delta),           if !isRightIncluded
+             *      kr = floor(right / delta) + 1,      if isRightIncluded
+             * The double division rd = right / delta never overflows and is
+             * exact, except in the presence of underflows. But even underflows
+             * do not affect the outcomes of ceil() and floor(), except,
+             * in turn, when the result drops to 0, that is, rd = 0.
+             *
+             * crd is a corrected version of rd when rd is zero. It is simply
+             * right / delta, but rounded away from 0 to preserve information
+             * ensuring correct outcomes in ceil() and floor().
+             *
+             * We know that -2^P <= kl, so
+             *      -2^P <= kl + nextLong(n)
+             * Also, since right <= -left, we know that
+             *      kr <= -kl + 1
+             * so that
+             *      0 < n <= -2 kl + 1
+             * This implies
+             *      kl + nextLong(n) <= kl + (-2 kl) = -kl <= 2^P
+             * and thus
+             *      -2^P <= kl + nextLong(n) <= 2^P
+             * which shows that kl + nextLong(n) can be cast exactly to double.
+             *
+             * Further, if isLeftIncluded then left = kl delta, so that we get
+             *      left = kl * delta <= (kl + nextLong(n)) * delta
+             * For any other k < kl, when nextLong(n) = 0 we would have
+             *      (k + nextLong(n)) * delta < left
+             * Otherwise, left = (kl - 1) delta, and therefore
+             *      left = (kl - 1) * delta < (kl + nextLong(n)) * delta
+             * For any other k < kl, when nextLong(n) = 0 we would get
+             *      (k + nextLong(n)) * delta <= left
+             * Either way, the lhs expression would not belong to I.
+             * That is, kl is the smallest integer such that kl delta always
+             * lies in I (it is an exact double).
+             *
+             * Similar considerations show that kr is the smallest integer such
+             * that kr delta lies to the right of I (it is an exact double).
+             *
+             * All the above means that (kl + nextLong(n)) * delta is an exact
+             * double lying in I and that kl and kr, thus n, are the best
+             * possible choices to ensure the largest number of equidistant
+             * doubles in I. Uniform distribution relies on the guarantee
+             * afforded by nextLong().
+             */
+            delta = nextUp(left) - left;
```

Comment:
The delta computations are equivalent to some expression phrased in terms of an ulp (Math.ulp) of a the endpoint or a value adjacent to the endpoint. If the ulp method is not used for the computation, I suggest adding a comment noting the equivalence.

Reply:
Here `left` is negative or `0.0`. What is needed is the distance to the next adjacent `double` in the direction of positive infinity. This is almost always the same as `Math.ulp(left)`, but not always.
For example, when `left` is `-1.0`, `delta` as computed here is half of `Math.ulp(left)`.

Analogously for the case covered by the `else` arm.